### PR TITLE
Loki: Higlight search words in queries when using template variables

### LINKED ISSUE #77861: Loki: Search Capability on RAW log lines (Using variable)
### What happened?

Datasource: Loki
Panel: logs

I am using loki data source and fetching RAW system logs.
in dashboard I have search bar to search some string/number in RAW logs.

image

I have created a "search" variable in the dashboard as a text box. and using this variable in a query like this:
“(?i)$search”

In previous Grafana version, it was highlighting the string/number searched in RAW logs, but now it does not.

When I search some number for result code, it does not highlight the text.
image

But instead of using variable in a query, If I directly put the text replacing "(?i)$search", it does highlight the text.
Example: {...} | "5002" instead of {...} | "(?i)$search"
image

In any of the above method, it should highlight the searched text, this is the only appropriate way to look at some text on the fly.
Without the text getting highlighted, even after searching for something in bunch of log lines you need to search for the text explicitly. 

### What did you expect to happen?

In any of the above method, it should highlight the searched text, this is the only appropriate way to look at some text on the fly.
Without the text getting highlighted, even after searching for something in bunch of log lines you need to search for the text explicitly. 

### Did this work before?

Yes, I guess in previous versions

### How do we reproduce it?

Mentioned in the "what happened"

### Is the bug inside a dashboard panel?

Maybe

### Environment (with versions)?

Grafana: v9.4.3
OS: Linux
Browser: Chrome

### Grafana platform?

Kubernetes

### Datasource(s)?

Loki

### PR DESCRIPTION
**Pull request by matyax**
This PR extends the search or highlighted words to include queries with template variables. Demo:

https://github.com/user-attachments/assets/11d36fb0-61fe-41b4-ab03-31b2fda0cf17

**What is this feature?**

Improvement over an existing functionality.

**Why do we need this feature?**

Feature request.

**Who is this feature for?**

Loki dashboard users.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/77861

**Special notes for your reviewer:**

Steps:
- Create a dashboard with textbox variables
- Create a query including said variable as a line filter
- Expected result: when the results include the string from the variable, it should be highlighted

### CODE REVIEW

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 1]
```diff
@@ -76,6 +77,23 @@ export function getHighlighterExpressionsFromQuery(input = ''): string[] {
   return results;
 }
 
+/**
+ * Returns search terms from a LogQL query with variables.
+ * Given a variable name $search with value foo:
+ * E.g., `{} |= $search |= "bar" != "baz"` returns `['foo']`.
```
Comment (svennergr) [inline public/app/plugins/datasource/loki/queryUtils.ts:83]:
```suggestion
 * E.g., `{} |= $search |= "bar" != "baz"` returns `['foo', 'bar']`.
```

Right?

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 2]
```diff
@@ -76,6 +77,23 @@ export function getHighlighterExpressionsFromQuery(input = ''): string[] {
...
+export function getVariableHighlighterExpressionsFromQuery(expr: string, templateSrv: TemplateSrv) {
+  const searchWords: string[] = [];
+
+  templateSrv.getVariables().forEach((variable) => {
+    if (templateSrv.containsTemplate(expr) && expr.includes(`$${variable.name}`)) {
+      searchWords.push(templateSrv.replace(`$${variable.name}`));
+    }
+  });
+
+  return searchWords;
```
Comment (svennergr) [inline public/app/plugins/datasource/loki/queryUtils.ts:94]: This is a bit tricky and not the usual way how to replace variables. 
There is a bunch of ways how users can write variables. E.g. `$var` or `${var}` or `${var:raw} is all valid.

We should use a method like `interpolateVariablesInQueries`, but then we do need access to `scopedVars`. I was thinking, if we should rather use the `response.schema.meta.executedQueryString` property that has the variables already replaced (after string manipulation): 
image

File: public/app/plugins/datasource/loki/queryUtils.ts (hunk 1 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/queryUtils.ts:83]: No. This function works in combination with the existing one, which combines both values coming from variables and those identified in the query.

Comment (svennergr) [reply public/app/plugins/datasource/loki/queryUtils.ts:83]: That's why it should return the value of `$search` and the value of `bar`, imo.

File: public/app/plugins/datasource/loki/queryUtils.ts (hunk 2 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/queryUtils.ts]: Done. Let me know what you think. Great idea, btw.

**Review by svennergr — APPROVED** [review]
Very cool! Nice!

File: public/app/plugins/datasource/loki/backendResultTransformer.test.ts [hunk 3]
```diff
@@ -232,4 +232,24 @@ describe('loki backendResultTransformer', () => {
     );
     expect(result.error?.message).toBe('parse error at line 1, col 2: invalid char escape');
   });
+
+  it('resolves search words from queries with template variables', () => {
+    const dataFrame = cloneDeep(inputFrame);
+    dataFrame.meta = {
+      executedQueryString: '{service_name="tns-app"} |~ "(?i)template" |= "variable"',
```
[the remark this answers is not in the document]
Comment (matyax) [reply public/app/plugins/datasource/loki/backendResultTransformer.test.ts:239]: Oh yes.

**Comment by sunilvishvakarma** [comment]
we are using the bitnami/grafana helm charts, so when can we expert the changes over there?

**Comment by matyax** [comment]
@sunilvishvakarma The changes are targeted for Grafana 11.3 which will be released on October 22nd.

**Pull request merged**